### PR TITLE
Update SubblockIndexProvider to not inherit from GeneralUserObject

### LINKED ISSUE #19328: Abstracting_SubblockIndexProvider
## Reason
Convert `SubblockIndexProvider` so that it a true abstract class and does not inherit from `GeneralUserObject`

## Design
Remove `GeneralUserObject` inheritance, change classes that so that they inherit from `SubblockIndexProvider` and from `ElementIntegralUserObject` as well.

## Impact
Once inheritance has been changed, no noticeable impact should be detected

### PR DESCRIPTION
**Pull request by vanwdani**
Closes #19328

### CODE REVIEW

File: modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #include "SubblockIndexProvider.h"
 #include "ADRankTwoTensorForward.h"
 #include "ADRankFourTensorForward.h"
+#include "ElementIntegralUserObject.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h:17]: This addition does not seem necessary

File: modules/tensor_mechanics/include/userobjects/SubblockIndexProvider.h [hunk 2]
```diff
@@ -9,24 +9,23 @@
...
-  static InputParameters validParams();
-
-  SubblockIndexProvider(const InputParameters & params) : GeneralUserObject(params) {}
-
+  virtual ~SubblockIndexProvider(){};
   /**
    * The index of subblock this element is on.
    */
-  virtual unsigned int getSubblockIndex(const Elem & /* elem */) const = 0;
+  virtual unsigned int getSubblockIndex(const libMesh::Elem & /* elem */) const = 0;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/userobjects/SubblockIndexProvider.h:28]: Yeah, this is not wrong, but we have a `using libMesh;` way uo in `Moose.h` so we use just `Elem` everywhere else in the code.

File: modules/tensor_mechanics/src/userobjects/SubblockIndexProvider.C [hunk 3]
```diff
@@ -8,9 +8,3 @@
 //* https://www.gnu.org/licenses/lgpl-2.1.html
 
 #include "SubblockIndexProvider.h"
-
```
Comment (dschwen) [inline modules/tensor_mechanics/src/userobjects/SubblockIndexProvider.C:11]: this file can be removed entirely

File: modules/tensor_mechanics/test/include/userobjects/TestSubblockIndexProvider.h [hunk 4]
```diff
@@ -10,13 +10,14 @@
...
 #include "SubblockIndexProvider.h"
+#include "ElementIntegralUserObject.h"
 
 /**
  * A class used to set the subblock index for testing generalized plane strain
  * calculations when more than one out-of-plane strain is provided on different
  * subsets of elements.
  */
-class TestSubblockIndexProvider : public SubblockIndexProvider
+class TestSubblockIndexProvider : public SubblockIndexProvider, public ElementIntegralUserObject
```
Comment (dschwen) [inline modules/tensor_mechanics/test/include/userobjects/TestSubblockIndexProvider.h:20]: This used to be a `GeneralUserObject` before, so why make it an `ElementIntegralUserObject`?

File: modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C [hunk 5]
```diff
@@ -14,13 +14,13 @@ registerMooseObject("TensorMechanicsTestApp", TestSubblockIndexProvider);
 InputParameters
 TestSubblockIndexProvider::validParams()
 {
-  InputParameters params = SubblockIndexProvider::validParams();
+  InputParameters params = ElementIntegralUserObject::validParams();
```
Comment (dschwen) [inline modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C:17]: just make it `InputParameters::emptyParameters()`

File: modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C [hunk 6]
```diff
@@ -14,13 +14,13 @@ registerMooseObject("TensorMechanicsTestApp", TestSubblockIndexProvider);
...
   params.set<ExecFlagEnum>("execute_on") = EXEC_INITIAL;
   return params;
 }
 
 TestSubblockIndexProvider::TestSubblockIndexProvider(const InputParameters & parameters)
-  : SubblockIndexProvider(parameters)
+  : ElementIntegralUserObject(parameters)
```
Comment (dschwen) [inline modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C:23]: revert this change

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h [hunk 7]
```diff
@@ -47,7 +50,7 @@ class GeneralizedPlaneStrainOffDiag : public DerivativeMaterialInterface<Kernel>
   unsigned int _scalar_out_of_plane_strain_var;
 
   /// A Userobject that carries the subblock ID for all elements
-  const SubblockIndexProvider * const _subblock_id_provider;
+  const SubblockIndexProvider * _subblock_id_provider;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h:53]: revert this change

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 8]
```diff
@@ -55,16 +56,20 @@ GeneralizedPlaneStrainOffDiag::GeneralizedPlaneStrainOffDiag(const InputParamete
...
     _num_disp_var(getParam<std::vector<VariableName>>("displacements").size()),
     _scalar_out_of_plane_strain_direction(getParam<MooseEnum>("out_of_plane_direction"))
 {
+  if (isParamValid("subblock_index_provider"))
+  {
+    const UserObject * uo = &(
+        _fe_problem.getUserObject<UserObject>(getParam<UserObjectName>("subblock_index_provider")));
+    _subblock_id_provider = dynamic_cast<const SubblockIndexProvider *>(uo);
+  }
+
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C:72]:
```suggestion
```

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 9]
```diff
@@ -55,16 +56,20 @@ GeneralizedPlaneStrainOffDiag::GeneralizedPlaneStrainOffDiag(const InputParamete
     _eigenstrain_names(getParam<std::vector<MaterialPropertyName>>("eigenstrain_names")),
     _deigenstrain_dT(_eigenstrain_names.size()),
     _scalar_out_of_plane_strain_var(coupledScalar("scalar_out_of_plane_strain")),
-    _subblock_id_provider(isParamValid("subblock_index_provider")
-                              ? &getUserObject<SubblockIndexProvider>("subblock_index_provider")
-                              : nullptr),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C:60]: revert this change

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 10]
```diff
@@ -16,6 +16,7 @@
 #include "MooseVariableScalar.h"
 #include "RankTwoTensor.h"
 #include "RankFourTensor.h"
+#include "UserObject.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C:19]: revert this change

File: modules/tensor_mechanics/src/userobjects/GeneralizedPlaneStrainUserObject.C [hunk 11]
```diff
@@ -81,7 +82,11 @@ void
 GeneralizedPlaneStrainUserObject::initialize()
 {
   if (isParamValid("subblock_index_provider"))
-    _subblock_id_provider = &getUserObject<SubblockIndexProvider>("subblock_index_provider");
+  {
+    const UserObject * uo = &(
+        _fe_problem.getUserObject<UserObject>(getParam<UserObjectName>("subblock_index_provider")));
+    _subblock_id_provider = dynamic_cast<const SubblockIndexProvider *>(uo);
+  }
```
Comment (dschwen) [inline modules/tensor_mechanics/src/userobjects/GeneralizedPlaneStrainUserObject.C:89]: revert

File: modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C [hunk 12]
```diff
@@ -8,20 +8,30 @@
...
 
 registerMooseObject("TensorMechanicsTestApp", TestSubblockIndexProvider);
 
 InputParameters
 TestSubblockIndexProvider::validParams()
 {
-  InputParameters params = SubblockIndexProvider::validParams();
+  InputParameters params = GeneralUserObject::validParams();
+  params.addParam<UserObjectName>("subblock_index_provider",
+                                  "SubblockIndexProvider user object name");
```
Comment (dschwen) [inline modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C:22]:
```suggestion
  InputParameters params = GeneralUserObject::validParams(); 
```

File: modules/tensor_mechanics/test/include/userobjects/TestSubblockIndexProvider.h [hunk 13]
```diff
@@ -36,4 +35,8 @@ class TestSubblockIndexProvider : public SubblockIndexProvider
    * The max index of subblock.
    */
   virtual unsigned int getMaxSubblockIndex() const override;
+
+protected:
+  /// A Userobject that carries the subblock ID for all elements
+  const SubblockIndexProvider * _subblock_id_provider;
```
Comment (dschwen) [inline modules/tensor_mechanics/test/include/userobjects/TestSubblockIndexProvider.h:41]: I don't understand why you add this. This object _is_ the SubblockIndexProvider, it doesn't couple to one

File: modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C [hunk 14]
```diff
@@ -8,20 +8,30 @@
...
-TestSubblockIndexProvider::TestSubblockIndexProvider(const InputParameters & parameters)
-  : SubblockIndexProvider(parameters)
+TestSubblockIndexProvider::TestSubblockIndexProvider(const InputParameters & params)
+  : GeneralUserObject(params)
 {
+  if (isParamValid("subblock_index_provider"))
+  {
+    const UserObject * uo = &(_fe_problem.getUserObject<UserObject>("subblock_index_provider"));
+    _subblock_id_provider = dynamic_cast<const SubblockIndexProvider *>(uo);
+  }
```
Comment (dschwen) [inline modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C:34]: remove

**Review by dschwen — DISMISSED** [review]

**Comment by dschwen** [comment]
I think what was tripping you up here was that you somehow started using `_fe_problem.getUserObject` rather than `getUserObject` (which is supplied by UserObjectInterface, and which does what I explained - take an InputParameter name).

File: modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C (hunk 14 above)
Comment (dschwen) [reply modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C:34]: not resolved

Comment (vanwdani) [reply modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C]: Needed to push still!

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 15]
```diff
@@ -65,6 +66,7 @@ GeneralizedPlaneStrainOffDiag::GeneralizedPlaneStrainOffDiag(const InputParamete
     _num_disp_var(getParam<std::vector<VariableName>>("displacements").size()),
     _scalar_out_of_plane_strain_direction(getParam<MooseEnum>("out_of_plane_direction"))
 {
+
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C:69]:
```suggestion
```

File: modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C [hunk 16]
```diff
@@ -8,19 +8,22 @@
 //* https://www.gnu.org/licenses/lgpl-2.1.html
 
 #include "TestSubblockIndexProvider.h"
+#include "Material.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/test/src/userobjects/TestSubblockIndexProvider.C:11]: material?

File: modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h [hunk 17]
```diff
@@ -15,6 +15,9 @@
 #include "ADRankTwoTensorForward.h"
 #include "ADRankFourTensorForward.h"
 
+// Forward Declarations
+class SubblockIndexProvider;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h:19]: You can remove this, as the header is included above

File: modules/tensor_mechanics/include/userobjects/SubblockIndexProvider.h [hunk 18]
```diff
@@ -8,21 +8,23 @@
...
 
 #pragma once
+#include "libmesh/elem.h"
 
-#include "GeneralUserObject.h"
+using libMesh::Elem;
 
+namespace libMesh
+{
+class Elem;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/userobjects/SubblockIndexProvider.h:17]: Same here. No need for the forward declaration if you're including the header

File: modules/tensor_mechanics/test/include/userobjects/TestSubblockIndexProvider.h [hunk 19]
```diff
@@ -10,13 +10,13 @@
 #pragma once
 
 #include "SubblockIndexProvider.h"
-
+#include "GeneralUserObject.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/test/include/userobjects/TestSubblockIndexProvider.h:13]: Empty line after

File: modules/tensor_mechanics/include/userobjects/SubblockIndexProvider.h [hunk 20]
```diff
@@ -8,21 +8,23 @@
...
  */
-class SubblockIndexProvider : public GeneralUserObject
+class SubblockIndexProvider
 {
 public:
-  static InputParameters validParams();
-
-  SubblockIndexProvider(const InputParameters & params) : GeneralUserObject(params) {}
-
+  virtual ~SubblockIndexProvider(){};
```
Comment (dschwen) [inline modules/tensor_mechanics/include/userobjects/SubblockIndexProvider.h:27]: Empty line after

**Review by dschwen — COMMENTED** [review]
Ok almost

**Comment by dschwen** [comment]
Let's just include `UserObject.h` from `UserObjectInterface.h`!

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 21]
```diff
@@ -16,6 +16,7 @@
 #include "MooseVariableScalar.h"
 #include "RankTwoTensor.h"
 #include "RankFourTensor.h"
+#include "UserObjectInterface.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C:19]: Kernel already derives from and includes user object interface. Is it really necessary to add this include here again?

**Comment by vanwdani** [comment]
@dschwen something with the non-unity only likes having an established path to call `getUserObject`.  For the classes with `#include Material.h`, I've used `_fe_problem.getUserObject` as seen in other classes, and it works (since `class UserObject` is forward defined in `Material.h`).  I've tried using `#include UserObjectInterface.h` and some other methods, but they don't seem to work.  Any insight on this?

**Comment by dschwen** [comment]
Do not use _fe_problem.getUserObject at all, ever. The issue must be
something else, but I'm on PL now :-)

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C (hunk 10 above)
Comment (recuero) [reply modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: @vanwdani, Don't we need this header to define the `UserObject` type in order to call `getUserObject<>` (not `_fe_problem.getUserObject`)? `UserObjectInterface.h` might not be enough for non-unity builds.

**Comment by dschwen** [comment]
That's why I suggested that Users object interface should include that
header.

**Comment by recuero** [comment]
That looks like a circular dependency.

**Comment by dschwen** [comment]
Pragma once will prevent circular dependencies. Including it ensures it is
available the first time it is required.

**Comment by vanwdani** [comment]
@loganharbour shall we try to merge this and the corresponding Bison MR?

**Review by loganharbour — APPROVED** [review]
BISON patch is ready, Okami patch will come after BISON patch. Merging despite failure
